LibIPC: Introduce async functionality to IPC - part 1

theIDinside opened the pull request:
## LibIPC: Introduce async functionality to IPC - part 1
Async IPC, the `internal Promise-implementation-detail`-variant

### Description:
This patch is part 1 in the draft for asynchronous IPC. This patch is
responsible for 100% for the functionality that is added to make async
IPC possible. The following patches in this series, contains the changes
to code in Ladybird at large, that needs to change in order to compile
and although I know this breaks with convention, I figured this would be
easier to review as a draft. I can squash/refactor, as soon as some
review of this has started. It also allows me to write additional
patches that can show how this should work.

### Changes made
Added a sequence id to all messages, which is provided by the endpoint
as a static monotonic id.

Changed the int value for messages and responses (for debuggability
purposes). All messages are uneven values and all responses are that
value + 1, to encode additional data into the value.

Messages that respond with data, now must be changed from:

`Messages::SomeEndpoint::ResponseType foo(Foo arg1, Bar arg2)`
to
`NonnullRefPtr<Core::Promise<Messages::SomeEndpoint::ResponseType>> foo(Foo arg1, Bar arg2)`

After in-depth discussion, maintainers of the project opted to go for
non-industry standard approach to maintain a logical cohesiveness with
ladybird, where the return of a Promise<T> is more logically suitable,
even if it incurs a little extra verbosity.

Existing IPC methods, now needs to do the following 4 things, and commit
2 of this Async IPC proposal will do a few of them for existing
messages:

- Change the return type from
  `Messages::EndPointName::MessageNameResponse` to
  `NonnullRefPtr<Messages::EndPointName::MessageName::Promise>`
- Construct the correct Promise type, which can be found in the
`Messages::EndPointName::MessageName::Promise` typedef, where `EndPointName`
and `Message` should be accordingly.
- Resolve the promise with a value, where appropriate.
- Return the promise

(Note: Under no circumstances are you allowed to configure a
`when_resolved` for this `Promise` as the IPC system relies on using
this to pickle the message and then write the message buffer over the
socket. You are only allowed to to resolve it with a value).

trflynn89 commented on Libraries/LibIPC/Connection.h:
```diff
@@ -81,38 +94,55 @@ class Connection : public ConnectionBase {
+        auto const msg = RequestType { LocalEndpoint::next_ipc_request_id(), forward<Args>(args)... };
+        MUST(post_message(msg));
+        m_resolvers.empend(msg.ipc_request_id(), [promise, msg_id = msg.message_id()](OwnPtr<IPC::Message> msg) {
+            ASSERT(msg->message_id() == msg_id + 1 && msg->endpoint_magic() == PeerEndpoint::static_magic());
```
We tend to reserve `ASSERT` for performance-critical code. I don't think this assertion (and a bunch below) are going to cause any perf issues, so to be safer, we'd stick with VERIFY here.

We would also break each of these into their own assertions, so if it breaks, we know which one:
```c++
VERIFY(msg->message_id() == msg_id + 1);
VERIFY(msg->endpoint_magic() == PeerEndpoint::static_magic());
```

trflynn89 commented on Libraries/LibIPC/Connection.h:
```diff
@@ -126,5 +156,4 @@ class Connection : public ConnectionBase {
         return nullptr;
     }
 };
-
```
Please keep this newline

trflynn89 commented on Libraries/LibIPC/Message.h:
```diff
@@ -70,9 +96,16 @@ class Message {
+    }
 
+private:
+    u64 m_request_id;
```
Initialize plain types
(Yes, we do this even if there's a constructor to initialize it).

trflynn89 commented on Libraries/LibIPC/Message.h:
```diff
@@ -70,9 +96,16 @@ class Message {
 
+private:
+    u64 m_request_id;
+};
```
newline here as well

trflynn89 commented on Services/WebContent/WebContentServer.ipc:
```diff
@@ -1,3 +1,4 @@
+#include "LibGfx/Color.h"
```
We use angle bracket includes

trflynn89 commented on Libraries/LibIPC/Message.cpp:
```diff
@@ -59,4 +59,9 @@ ErrorOr<void> MessageBuffer::transfer_message(Transport& transport)
     return {};
 }
 
+/* static */
```
not necessary

trflynn89 commented on Libraries/LibWebView/BrowserProcess.h:
```diff
@@ -19,6 +20,8 @@
 
 namespace WebView {
 
+using namespace Messages::UIProcessServer;
```
Please no

ADKaster replied:
To expand on this, `using namespace` in header files is strictly verboten. And in cpp files it's discouraged, with a few specific exceptions (AK::TimeLiterals for example).

trflynn89 commented on Libraries/LibWebView/BrowserProcess.h:
```diff
@@ -6,6 +6,7 @@
 
 #pragma once
 
+#include "LibIPC/Stub.h"
```
We use angle bracket includes
(I'll not comment on the remainder of these)

trflynn89 commented on Libraries/LibWebView/WebContentClient.cpp:
```diff
@@ -470,24 +471,25 @@ void WebContentClient::did_change_favicon(u64 page_id, Gfx::ShareableBitmap favi
+void WebContentClient::did_set_cookie(URL::URL url, Web::Cookie::ParsedCookie cookie, Web::Cookie::Source source, DidSetCookie::Resolver resolve)
 {
     Application::cookie_jar().set_cookie(url, cookie, source);
+    return resolve();
```
`return`? This is a `void` function

Hendiadyoin1 replied:
(SideNote: There is one case where a "valued" void return makes sense, and that is enforcing tail calls with compiler attributes)

theIDinside replied:
> (SideNote: There is one case where a "valued" void return makes sense, and that is enforcing tail calls with compiler attributes)

Thank you for your input. The parts involving actual IPC message implementations are just to show how the new system would work - and so I did a sort of transpiling of the implementations, I myself did not know or understand why "returning" a void was needed.

theIDinside replied:
> `return`? This is a `void` function

This came as a part of a re-write of old implementations to show how it should be used. In this particular case doing the first scenario below works just fine, since there's no other branching etc going on after that. But both scenarios are useful:

```cpp
if(foo) {
  resolve();
  return;
}
```

can be spelled

```cpp
// When resolving, the function should also return
if(foo)
  return resolve();
```

Since the callbacks themselves are guaranteed to return void, this is fine to do.

theIDinside replied:
Btw, @ADKaster you wrote IPC messages that return void responses, see `UIProcessServer.ipc`: whether the resolve mechanism is promise or a `Resolver` an empty resolve needs to happen.

If what you wrote there is a bug, it should probably be fixed separately from this patch series, right?

trflynn89 commented on Libraries/LibWebView/WebContentClient.cpp:
```diff
@@ -500,14 +502,14 @@ void WebContentClient::did_expire_cookies_with_time_offset(AK::Duration offset)
     }
-
-    return String {};
+    return resolve(String{});
```
Here as well re: `return`, and many places below

theIDinside replied:
This was part of automated rewrite, in the final PR, these will be removed. `return resolve(value)` is as stated in other comment thread, useful for when we want to also exit the funciton - but as the last statement, it's just noise (as the function will return now anyway)

trflynn89 commented on Libraries/LibWebView/WebContentClient.cpp:
```diff
@@ -500,14 +502,14 @@ void WebContentClient::did_expire_cookies_with_time_offset(AK::Duration offset)
-            return view->on_new_web_view(activate_tab, hints, page_index);
+        if (view->on_new_web_view) {
+            return resolve(view->on_new_web_view(activate_tab, hints, page_index));
+        }
```
There's no need to add angle brackets to existing 1-line `if` statements

trflynn89 reviewed (commented):
I saw there was ongoing design discussion about this change on Discord, so for now, I've only non-exhaustively reviewed project coding style issues. Please read https://github.com/LadybirdBrowser/ladybird/tree/master/Documentation/CodingStyle.md and https://github.com/LadybirdBrowser/ladybird/blob/master/CONTRIBUTING.md.

For atomicity, the second commit should  really be multiple commits - one for each component being ported to this new infrastructure.

(This is why we recommend starting with smaller changes instead of jumping into a big architectural change. There's always lots of churn over project styles and practices.)

Having async IPC will be very exciting though :)

ADKaster commented on Libraries/LibIPC/Message.cpp:
```diff
@@ -59,4 +59,9 @@ ErrorOr<void> MessageBuffer::transfer_message(Transport& transport)
+/* static */
+RefPtr<MessageResolverState> MessageResolverState::create(u64 request_id)
+{
+    return RefPtr<MessageResolverState> { new MessageResolverState { request_id } };
```
this leaks immediately. you have to `adopt_ref`. c.f. https://github.com/LadybirdBrowser/ladybird/blob/master/Documentation/SmartPointers.md#manual-construction-1

ADKaster commented on Libraries/LibIPC/Connection.h:
```diff
@@ -81,38 +94,55 @@ class Connection : public ConnectionBase {
+    NonnullOwnPtr<typename RequestType::ResponseType> send_sync(Args&&... args)
     {
-        return wait_for_specific_endpoint_message<MessageType, LocalEndpoint>();
+        auto result = send<RequestType, Args...>(forward<Args>(args)...)->await();
```
this implementation of send_sync is behaviorally very different from the prior implementation. This implementation will add an event loop spin to the stack, and continue processing other events. `ConnectionBase::wait_for_specific_endpoint_message_impl` on master completely blocks execution by checking only the one specific socket in question, and causing any other events or IPC messages to be queued until the correct response type is received.

theIDinside replied:
Yes, and for deprecated XMLHttpRequest's (the sync api version) we will need to maintain the old "actually blocking sync", as was discussed on discord. So sync *really* must mean blocking-the-event-loop sync not the one I had initially thought was ok.

ADKaster commented on Libraries/LibWebView/BrowserProcess.cpp:
```diff
@@ -114,16 +114,20 @@ void UIProcessConnectionFromClient::die()
 }
 
-void UIProcessConnectionFromClient::create_new_window(Vector<ByteString> urls)
+void UIProcessConnectionFromClient::create_new_window(Vector<ByteString> urls, CreateNewWindow::Resolver resolve)
```
as the resolver is an object, its variable name should be a noun, not a verb. I checked the surrounding code to see if you had created a new member function called `resolve` when I first glanced at this, which is not good.

ADKaster commented on Services/WebContent/WebDriverConnection.cpp:
```diff
@@ -251,43 +254,59 @@ void WebDriverConnection::set_is_webdriver_active(bool is_webdriver_active)
 
     // 1. Let timeouts be the result of trying to JSON deserialize as a timeouts configuration the request’s parameters.
-    TRY(Web::WebDriver::json_deserialize_as_a_timeouts_configuration_into(payload, m_timeouts_configuration));
+    RESOLVE_IF_ERROR(Web::WebDriver::json_deserialize_as_a_timeouts_configuration_into(payload, m_timeouts_configuration));
```
eh.. resolve if error? I think this is normally spelled "reject".

ADKaster commented on Services/WebContent/WebDriverConnection.cpp:
```diff
@@ -251,43 +254,59 @@ void WebDriverConnection::set_is_webdriver_active(bool is_webdriver_active)
 }
 
+#define RESOLVE(...) resolve(Web::WebDriver::Response { __VA_ARGS__ })
+#define ER() resolve(Web::WebDriver::Response { JsonValue{} })
```
this macro name "ER" is far too short to be useful

theIDinside replied:
Absolutely, it was just for peace of mind when renaming all this stuff. We have this convention scattered throughout `WebDriver` where we return (serialize and write back) empty json values, only to at some point later, ad-hoc respond with the actual value.

In the final product as it were, all macros will be removed (because all functions will resolve using their respective `Resolver`).

ADKaster commented on Services/WebContent/WebDriverConnection.cpp:
```diff
@@ -1433,14 +1447,14 @@ Messages::WebDriverClient::GetElementTagNameResponse WebDriverConnection::get_el
 
 // 12.4.7 Get Element Rect, https://w3c.github.io/webdriver/#dfn-get-element-rect
-Messages::WebDriverClient::GetElementRectResponse WebDriverConnection::get_element_rect(String element_id)
+void WebDriverConnection::get_element_rect(String element_id, IPCMessages::GetElementRect::Resolver resolve)
```
I'm not sure I understand why, from an IPC connection perspective, it makes sense to have this argument? I might be missing something. But I don't see the difference in these call sites in this file between `return Blah` and `TRY(blah)` directly vs these extra macros you created to defer to an extra argument passed in. It seems strictly worse, tbh, because now all these things need to return void instead.

theIDinside replied:
It's better because it allows for the handler code to actually also be async. I'd argue that returning the value is much worse, as this now requires the handler code to be synchronous and you have to write all these ad-hoc ping-pong messages that we have scattered through out ladybird.

theIDinside replied:
This kind of stuff lends itself to be drawn generally. Excuse my poor handwriting.
![image](https://github.com/user-attachments/assets/0ae63ca1-a49f-4fdc-96b0-d09c25a84836)

Where as in the return-case, the one that enforces WebDriver to return empty json values, it looks like this

![image](https://github.com/user-attachments/assets/462aa85c-0a7f-4141-a6fc-356958260331)

And that's essentially what we want to avoid, using resolvers.

ADKaster replied:
I... Suppose, but why aren't we just returning a `NNRP<Core::Promise>`?

theIDinside replied:
Yeah, maybe that wasn't clear; because A and B are two different processes. What the resolver function does is take the parameters, serialize that date and writes the response.

alimpfard replied:
Re Promise, it's basically the same, with different ownership - in the case of your resolver function, the function owns a "handle" and pushes the data through that when called
With a Core::Promise, the caller hangs the "pushes the data through handle" on the promise object's on_resolved which gets called when it's...resolved.

alimpfard replied:
i.e. instead of the caller providing a function to be called when the value is ready, the callee produces a promise that gets resolved when the value is ready, and the caller can do whatever with it.

theIDinside replied:
> i.e. instead of the caller providing a function to be called when the value is ready, the callee produces a promise that gets resolved when the value is ready, and the caller can do whatever with it.

That is exactly how it works for `A`, yes.

theIDinside replied:
The one who handles the message, to provide the response shall not return a Promise. There the caller of the handler function is internal code, produced by the IPC compiler. All it wants to do, is write that data over the socket back to A.

Using a promise there has no use cases, because this is not code exposed for users to see or use. It's internal to the IPC system, just as how the previously returned value, was internal to the IPC system (because it immediately gets serialized + written over the socket).

theIDinside replied:
Here's some code produced by the IPC compiler right now, for `RequestServerStub`:

```cpp
    virtual ErrorOr<OwnPtr<IPC::MessageBuffer>> handle(NonnullOwnPtr<IPC::Message> message) override
    {
        switch (message->message_id()) {

        case (int)Messages::RequestServer::MessageID::InitTransport: {
            [[maybe_unused]] auto& request = static_cast<Messages::RequestServer::InitTransport&>(*message);
            auto response = init_transport(request.peer_pid());
            // This value, is returned to `ConnectionBase`, who immediately writes the data to socket
            return make<IPC::MessageBuffer>(TRY(response.encode()));
        }
```

As we can see here,  `init_transport` is required to always be synchronous, because it's returning a value. This is the reason why we want to provide a Resolver (or a Promise, any asynchronous operation really), is because it's type checked, and the end user (the one who) when it wants to respond, says `resolve(a, b, ,c)` which can be at any time in the future, it doesn't have to be immediately.

The question becomes, is it preferable for the implementers of IPC messages to involve themselves with the guts of Promises, how to construct them, how to resolve them? Because if that's the case, sure, we can do that. Then the one who calls `RequestServerStub::handle` will have to manage promise's (and their life times)  and hook up a Resolver-like function for it's `when_resolved`.

I think this is a worse interface for the user, than having it simply say `resolve(a, b, c)` where `a, b, c` are the parameters of the `ResponseType` for the message. Whereas for the `Promise` approach, the user would have to first
- construct a promise
- construct the right response type, move the parameters to make sure that we're not doing any copies (whereas `resolve` takes const&, because it's *sole job* is to serialize + write and so can do that directly)
- resolve the promise, with that value, upon which `when_resolved` is immediately called and serializes + writes the data

theIDinside replied:
And this is how `ImageDecoderServerEndpoint.h` looks like (for the `InitTransport` message, which responds with a  `InitTransportResponse`), under the new ipc compiler changes (and the new async ipc), 

```cpp
template <typename ResponseType, typename ...ResponseArguments>
static auto create_resolver(NonnullRefPtr<IPC::ConnectionBase> conn, u64 seq_id) {
    return [state=IPC::MessageResolverState::create(seq_id), conn](ResponseArguments&&... args) {
        VERIFY(state->can_resolve());
        state->set_resolved();
        if constexpr(sizeof...(ResponseArguments) == 0) {
            auto res = ResponseType::static_encode(state->request_id());
            if (auto const post_result = conn->post_message(res.release_value()); post_result.is_error()) {
                dbgln("IPC::ConnectionBase::handle_messages: {}", post_result.error());
            }
        } else {
            auto res = ResponseType::static_encode(state->request_id(), std::forward<ResponseArguments>(args)...);
            if (auto const post_result = conn->post_message(res.release_value()); post_result.is_error()) {
                dbgln("IPC::ConnectionBase::handle_messages: {}", post_result.error());
            }
        }
    };
}

virtual void handle_ipc_message([[maybe_unused]] NonnullRefPtr<IPC::ConnectionBase> conn, NonnullOwnPtr<IPC::Message> message) override
{
    switch (message->message_id()) {

    case (int)Messages::ImageDecoderServer::MessageID::InitTransport: {
        [[maybe_unused]] auto& request = static_cast<Messages::ImageDecoderServer::InitTransport&>(*message);
        // a resolver is created here, that can post the results back to A
        return init_transport(request.peer_pid(), create_resolver<Messages::ImageDecoderServer::InitTransportResponse, const int&>(std::move(conn), message->ipc_request_id()));
    }
        ```

alimpfard replied:
> Using a promise there has no use cases, because this is not code exposed for users to see or use. It's internal to the IPC system, just as how the previously returned value, was internal to the IPC system (because it immediately gets serialized + written over the socket).

No, but the function returning that promise is hand-written, and ergonomics matters there.
Yes the function consuming that promise has no need for it to be a promise, but the function returning that promise...does.

ADKaster replied:
I still don't understand. why is it a problem for the implementer of MyConnectionFromServer to construct a promise?

```
virtual void handle_ipc_message([[maybe_unused]] NonnullRefPtr<IPC::ConnectionBase> conn, NonnullOwnPtr<IPC::Message> message) override
{
    switch (message->message_id()) {

    case (int)Messages::ImageDecoderServer::MessageID::InitTransport: {
        [[maybe_unused]] auto& request = static_cast<Messages::ImageDecoderServer::InitTransport&>(*message);
        auto promise = init_transport(request.peer_pid());
        promise.when_resolved = [id = message->ipc_request_id()](auto... args) {
            auto response = ResponseType::encode(id, forward<decltype(args)>(args))
            post_message(result);
        };
    }
```

Why do we need to invent this extra bespoke pseudo-promise pattern?

We could even have the promise type be the `ResponseType` guy, and just have a setter for the request id that only LibIPC knows about.

If you're worried about the lifetime of the promise, that's what that event loop add_job thingy is for.

theIDinside replied:
Oh it's not a problem of course, you raised an issue about what's worse so I was just discussing what's worse and the promise approach certainly is less ergonomic and worse than the Promise approach.

Here follows examples of the different implementations for `WebDriverConnection::find_element_from_element` and if you still think Promise is better, I can certainly implement that pattern. I mean I won't enjoy it, but I'll do it.

```cpp
// 12.3.4 Find Element From Element, https://w3c.github.io/webdriver/#dfn-find-element-from-element
Messages::WebDriverClient::FindElementFromElementResponse WebDriverConnection::find_element_from_element(JsonValue payload, String element_id)
{
  // >>> Now, the user must construct it's own promise, where as for the ergonomic version we pass a `Resolver`,
  // >>> sort of like Mojo, sort of like Firefox's IPDL. In both cases we're helped by the IPC compiler to
  // >>> produce a typedef, so it doesn't get that much more noisy in the Promise scenario:
  auto promise = Promise<Messages::WebDriverClient::FindElementFromElementResponse>::construct();
  // >>> We could also let the IPC compiler produce a typedef for the exact template instantiation
  // >>>  auto promise = Messages::WebDriverClient::FullscreenWindowResponsePromise::construct();

  // 1. Let location strategy be the result of getting a property named "using" from parameters.
  // >>> Macros like TRY, needs re-writing in both cases, so there's no win/loss here for either implementation.
  auto location_strategy_string = TRY(Web::WebDriver::get_property(payload, "using"sv));
  auto location_strategy = Web::WebDriver::location_strategy_from_string(location_strategy_string);
  // 2. If location strategy is not present as a keyword in the table of location strategies, return error with error code invalid argument.
  if (!location_strategy.has_value())
    return Web::WebDriver::Error::from_code(Web::WebDriver::ErrorCode::InvalidArgument, MUST(String::formatted("Location strategy '{}' is invalid", location_strategy_string)));
  // 3. Let selector be the result of getting a property named "value" from parameters.
  // 4. If selector is undefined, return error with error code invalid argument.
  auto selector = TRY(Web::WebDriver::get_property(payload, "value"sv));
  // 5. If session's current browsing context is no longer open, return error with error code no such window.

  TRY(ensure_current_browsing_context_is_open());
  // 6. Try to handle any user prompts with session.
  // >>> Both the `Promise` impl and the `Resolver` impl will require the end user to make sure that the right closure captures it.
  handle_any_user_prompts([this, element_id, location_strategy, selector = move(selector), promise]() mutable {
    auto get_start_node = GC::create_function(current_browsing_context().heap(), [this, element_id]() -> ErrorOr<GC::Ref<Web::DOM::ParentNode>, Web::WebDriver::Error> {
      // 7. Let start node be the result of trying to get a known element with session and URL variables["element id"].
      return Web::WebDriver::get_known_element(current_browsing_context(), element_id);
    });
    // 8. Let result be the value of trying to Find with session, start node, location strategy, and selector.
    // >>> Both the `Promise` impl and the `Resolver` impl will require the end user to make sure that the right closure captures it.
    find(*location_strategy, move(selector), get_start_node, GC::create_function(current_browsing_context().heap(), [this, promise](Web::WebDriver::Response result) {
      // 9. If result is empty, return error with error code no such element. Otherwise, return the first element of result.
      // >>> Now, we must constrcut Messages::WebDriverClient::FindElementFromElementResponse correctly,
      // >>> and this involves making the IPC Compiler produce constructors for the type, that knows whether the type is helped by moving/copying
      // >>> which is non-trivial, and only works currently as a best guess-effort. In this specific scenario, it's easy. Move everything.
      // >>> This is certainly not true for all response types. There can be many combinations of copies and moves.
      promise.resolve(Messages::WebDriverClient::FindElementFromElementResponse{extract_first_element(move(result))});
      // old implementation: async_driver_execution_complete(extract_first_element(move(result)));
    }));
  });
  // >>> Now we return the promise to ConnectionBase, which will hook up a when_resolved (hoping that the IPC message implementer doesn't
  // >>> do it, at some point in the future, overriding the behavior).
  // >>> That `when_resolved` will be an exact implementation of what the `Resolver` type does, token for token.
  return promise;
}
```

And here's the resolver implementation

```cpp// 12.3.4 Find Element From Element, https://w3c.github.io/webdriver/#dfn-find-element-from-element
Messages::WebDriverClient::FindElementFromElementResponse WebDriverConnection::find_element_from_element(JsonValue payload, String element_id, FindElementFromElementResponse::Resolver resolve)
{
  // >>> No construction of a promise. No requirements of intimate knowledge of the IPC system required. Just call resolve(a, b, c).

  // 1. Let location strategy be the result of getting a property named "using" from parameters.
  // >>> Macros like TRY, needs re-writing in both cases, so there's no win/loss here for either implementation.
  auto location_strategy_string = TRY(Web::WebDriver::get_property(payload, "using"sv));
  auto location_strategy = Web::WebDriver::location_strategy_from_string(location_strategy_string);
  // 2. If location strategy is not present as a keyword in the table of location strategies, return error with error code invalid argument.
  if (!location_strategy.has_value())
    return Web::WebDriver::Error::from_code(Web::WebDriver::ErrorCode::InvalidArgument, MUST(String::formatted("Location strategy '{}' is invalid", location_strategy_string)));
  // 3. Let selector be the result of getting a property named "value" from parameters.
  // 4. If selector is undefined, return error with error code invalid argument.
  auto selector = TRY(Web::WebDriver::get_property(payload, "value"sv));
  // 5. If session's current browsing context is no longer open, return error with error code no such window.

  TRY(ensure_current_browsing_context_is_open());
  // 6. Try to handle any user prompts with session.
  // >>> Both the `Promise` impl and the `Resolver` impl will require the end user to make sure that the right closure captures it.
  handle_any_user_prompts([this, element_id, location_strategy, selector = move(selector), resolve]() mutable {
    auto get_start_node = GC::create_function(current_browsing_context().heap(), [this, element_id]() -> ErrorOr<GC::Ref<Web::DOM::ParentNode>, Web::WebDriver::Error> {
      // 7. Let start node be the result of trying to get a known element with session and URL variables["element id"].
      return Web::WebDriver::get_known_element(current_browsing_context(), element_id);
    });
    // 8. Let result be the value of trying to Find with session, start node, location strategy, and selector.
    // >>> Both the `Promise` impl and the `Resolver` impl will require the end user to make sure that the right closure captures it.
    find(*location_strategy, move(selector), get_start_node, GC::create_function(current_browsing_context().heap(), [this, resolve](Web::WebDriver::Response result) {
      // 9. If result is empty, return error with error code no such element. Otherwise, return the first element of result.
      // >>> No construction for end user. Just call resolve(result)
      resolve(result);
      // old implementation: async_driver_execution_complete(extract_first_element(move(result)));
    }));
  });
  // >>> Return nothing. Because what we return, is supposed to be written to socket immediately anyway, or at
  // >>> least serialized directly.
}
```

So if we're talking ergonomics here, it should be apparent that the `Resolver` approach is superior to the Promise approach that places the following requirements on the end user:
- Construction of the response type (which is an internal detail, now being exposed to the end user)
- Not meddling with the promise's internal state in any fashion, except resolving the value
- While creating the response type, we now must involve ourselves with moves, etc construction an object for no reason at all, which adds unnecessary overhead - and that construction is entirely unnecessary!

Is this not less ergonomic than simply having `resolve(a, b, c)`? Because I don't see how `Promise`s would be more ergonomic @alimpfard ?

theIDinside replied:
I suppose another argument in favor of the `Resolve`r approach is much more common than the proposed `Promise` variant. Mojo and IPDL both work similarly (on the interface level, that is) as the `Resolve`r approach.

Here's a IPC message in Mojo:

```mojo
  GetTextSurroundingSelection(uint32 max_length)
      => (mojo_base.mojom.String16 content, uint32 start_offset,
          uint32 end_offset);
```

And the implementation in Chrome:

```
  // blink::mojom::LocalFrame overrides:
  void GetTextSurroundingSelection(
      uint32_t max_length,
      GetTextSurroundingSelectionCallback callback) final;
```

Which is then implemented like so:

```cpp
void LocalFrameMojoHandler::GetTextSurroundingSelection(
    uint32_t max_length,
    GetTextSurroundingSelectionCallback callback) {
  SurroundingText surrounding_text(frame_, max_length);

  if (surrounding_text.IsEmpty()) {
    // Don't use WTF::String's default constructor so that we make sure that we
    // always send a valid empty string over the wire instead of a null pointer.
    std::move(callback).Run(g_empty_string, 0, 0);
    return;
  }
  // Essentially their version of our `Resolve`
  std::move(callback).Run(surrounding_text.TextContent(),
                          surrounding_text.StartOffsetInTextContent(),
                          surrounding_text.EndOffsetInTextContent());
}
```

I don't think we should sacrifice performance (even if it's entirely negligible), and make less sense on the interface level than well established design patterns (used in libraries like Mojo and IPDL for instance), just to return `Promise`.

Remember; A `Promise<T>` is a design to signal that a value of `T` will some time in the future exist. It's used to signal an asynchronously produced value. That's not what B is doing in this scenario, B is dealing with *asynchronus operation* (the pickling/serialization and the write). It's on A's side, that requested the value, _there_ it makes perfect sense to use a `Promise` because there, it's expecting to actually be a value. And the proposal here, does do that.

alimpfard replied:
Looking at this in isolation isn't really meaningful imo, this codebase has patterns and a magic resolve function is just not in those; introducing new patterns has ergonomic costa regardless of how common they are elsewhere.

As for the macros, there's no need to introduce new macros, the function can always return ErrorOr<NNRP<Promise<...>>> (which makes more sense to me anyway, early errors don't need to be communicated "later').

I will refrain from talking about performance, but if you wish to make claims, please show measurements. Very rarely can one reason about it correctly on the level of abstract objects.

And here's a sample view into a possible future:
```c++
// 12.3.4 Find Element From Element, https://w3c.github.io/webdriver/#dfn-find-element-from-element
ErrorOr<NonnullRefPtr<Promise<Messages::WebDriverClient::FindElementFromElementResponse>> WebDriverConnection::find_element_from_element(JsonValue payload, String element_id)
{
  // >>> Now, the user must construct it's own promise, where as for the ergonomic version we pass a `Resolver`,
  auto promise = Promise<Messages::WebDriverClient::FindElementFromElementResponse>::construct();

  // 1. Let location strategy be the result of getting a property named "using" from parameters.
  auto location_strategy_string = TRY(Web::WebDriver::get_property(payload, "using"sv));
  auto location_strategy = Web::WebDriver::location_strategy_from_string(location_strategy_string);
  // 2. If location strategy is not present as a keyword in the table of location strategies, return error with error code invalid argument.
  if (!location_strategy.has_value()) {
    return Web::WebDriver::Error::from_code(Web::WebDriver::ErrorCode::InvalidArgument, MUST(String::formatted("Location strategy '{}' is invalid", location_strategy_string)));
  }
  // 3. Let selector be the result of getting a property named "value" from parameters.
  // 4. If selector is undefined, return error with error code invalid argument.
  auto selector = TRY(Web::WebDriver::get_property(payload, "value"sv));
  // 5. If session's current browsing context is no longer open, return error with error code no such window.

  TRY(ensure_current_browsing_context_is_open());
  // 6. Try to handle any user prompts with session.
  // >>> Both the `Promise` impl and the `Resolver` impl will require the end user to make sure that the right closure captures it.
  handle_any_user_prompts([this, element_id, location_strategy, selector = move(selector), promise]() mutable {
    auto get_start_node = GC::create_function(current_browsing_context().heap(), [this, element_id]() -> ErrorOr<GC::Ref<Web::DOM::ParentNode>, Web::WebDriver::Error> {
      // 7. Let start node be the result of trying to get a known element with session and URL variables["element id"].
      return Web::WebDriver::get_known_element(current_browsing_context(), element_id);
    });
    // 8. Let result be the value of trying to Find with session, start node, location strategy, and selector.
    // >>> Both the `Promise` impl and the `Resolver` impl will require the end user to make sure that the right closure captures it.
    find(*location_strategy, move(selector), get_start_node, GC::create_function(current_browsing_context().heap(), [this, promise](Web::WebDriver::Response result) {
      // 9. If result is empty, return error with error code no such element. Otherwise, return the first element of result.
      // >>> Now, we must constrcut Messages::WebDriverClient::FindElementFromElementResponse correctly,
      // >>> no need for explicit constructors, unless this type is not an aggregate?
      promise.resolve({extract_first_element(move(result))});
      // old implementation: async_driver_execution_complete(extract_first_element(move(result)));
    }));
  });
  // >>> Now we return the promise to ConnectionBase, which will hook up a when_resolved (hoping that the IPC message implementer doesn't
  // >>> do it, at some point in the future, overriding the behavior and breaking the IPC system).
  // >>> That `when_resolved` will be an exact implementation of what the `Resolver` type does, i.e. serialize + write
  return promise;
}

alimpfard replied:
Also of note, Andrew mentioned that we have a promise_try somewhere as well, so the function _could_ do that instead of ErrorOr<...>

theIDinside replied:
I guess we could write a Promise adapter that prevents an end user to misuse the promise (by overriding when_resolved at some time later making the IPC system blow up), or it should be caught in review whenever we see a `when_resolved` that this is disallowed.

ADKaster replied:
There's a lot of words in this thread now making it hard to follow, but I will say one thing. It's a lot more important to be consistent with patterns we are already using in our codebase than it is to copy an 'industry standard' pattern from elsewhere. That's not to say this resolver pattern is bad, it's just that it's unique to this PR from Ladybird's perspective. In a choice between re-using and existing pattern that solves most of the problems vs inventing a new tailor-made pattern, I'll choose pattern re-use every time.

The biggest motivation for me to push for a promise-based pattern is that we already use this pattern copiously in LibWeb for promise-returning Web APIs. The less cognitive load there is to move between libraries in the codebase, the better.

theIDinside replied:
> Also of note, Andrew mentioned that we have a promise_try somewhere as well, so the function _could_ do that instead of ErrorOr<...>

Right, I think that particular issue is not that big of a problem, but any facilities to reduce clutter obviously is welcomed usually.

theIDinside replied:
> There's a lot of words in this thread now making it hard to follow, but I will say one thing. It's a lot more important to be consistent with patterns we are already using in our codebase than it is to copy an 'industry standard' pattern from elsewhere. That's not to say this resolver pattern is bad, it's just that it's unique to this PR from Ladybird's perspective. In a choice between re-using and existing pattern that solves most of the problems vs inventing a new tailor-made pattern, I'll choose pattern re-use every time.

> The biggest motivation for me to push for a promise-based pattern is that we already use this pattern copiously in LibWeb for promise-returning Web APIs. The less cognitive load there is to move between libraries in the codebase, the better.

I can buy that, even if it's an anti-pattern in this case (since we're not actually providing a value asynchronously on B's side, what we want to do is an asynchronous operation. The asynchronously produced value is actually on A's side).

We'll have to also remember to never override `when_resolved` of the Promise inside the IPC function, as this will blow up the IPC system (since it requires `when_resolved` to be unset, and never be overridden either), so it will be required of reviewers going forward for all IPC messages that this can't be done.

Changing from the `Resolver` to the `Promise` variant, beyond that, is not really that difficult to change, just some small changes to the IPC Compiler, for instance produce the `ResponsePromiseType`, as well some very minor plumbing (the example you wrote for instance, with the `when_resolved`).

theIDinside replied:
> > it should be caught in review whenever we see a `when_resolved` that this is disallowed.
> 
> How would this happen exactly? Only the function making the promise is user-controlled, and adding a resolve handler to a promise that will be resolved somewhere in the current function (lexically) is not something I've seen - so it _would_ stand out.

It'd definitely stand out. But remember, the handler code does this:

```cpp
// this promise is returneed by the ipc handler method
auto promise = ipc_handler_method(...)
promise.when_resolved([]() {
   // pickle + write message to socket
});
```

But inside the IPC message handler, we may be dispatching async stuff, and if we set `when_resolved` for the promise in those, then we've overridden the `promise.when_resolved` set by the IPC Message handler. But like you said, this would stand out, and going forward, we would disallow this for all Promises returned from IPC messages.

theIDinside replied:
Got most of the IPC compiler re-work done, checking if this is what you would want instead;

```cpp
NonnullRefPtr<InitTransport::Promise> ConnectionFromClient::init_transport([[maybe_unused]] int peer_pid)
{
#ifdef AK_OS_WINDOWS
    m_transport.set_peer_pid(peer_pid);
    auto promise = Messages::ImageDecoderServer::InitTransport::Promise::construct();
    promise.resolve(Messages::ImageDecoderServer::InitTransportResponse{Core::System::getpid()});
    return promise;
#endif
    VERIFY_NOT_REACHED();
}
```

theIDinside commented:
> I saw there was ongoing design discussion about this change on Discord, so for now, I've only non-exhaustively reviewed project coding style issues. Please read https://github.com/LadybirdBrowser/ladybird/tree/master/Documentation/CodingStyle.md and https://github.com/LadybirdBrowser/ladybird/blob/master/CONTRIBUTING.md.
> 
> For atomicity, the second commit should really be multiple commits - one for each component being ported to this new infrastructure.
> 
> Be sure to run clang-format over your changes.
> 
> (This is why we recommend starting with smaller changes instead of jumping into a big architectural change. There's always lots of churn over project styles and practices.)
> 
> Having async IPC will be very exciting though :)

Thank you for your inputs, they are appreciated greatly!

I think I've caught most nits and major guidelines violations, I've double and triple checked, and hopefully my monkeybrain has not made any glaring mistakes!

The new patch series also addresses concerns by @ADKaster and @alimpfard that want to stay logically consistent with the ladybird codebase at large, though I'm sure they'll come to see it my way, some day :P. Either way, both solutions end up doing what's important here, providing asynchronous IPC so ladybird code does not have to write ad-hoc/bespoke communication for everything.

I think there can be a patch #3 to this series, where re-write of WebDriver (see commit message 2) which would be useful for future contributors and people who are unfamiliar with IPC things generally, as it would be a great place to show "this is how it's supposed to work" - but as stated in that commit message, it's not strictly necessary.

theIDinside commented:
I think the linter may have a bug. That second commit title is not longer than 72 (it's 53 characters long), but it has appended some hash to it in the error message (which does not exist in the actual commit title).
